src/DataStructures/DataBox/DataBox.hpp

```diff
@@ -1793,6 +1793,36 @@ inline constexpr auto mutate_apply(
                                       std::forward<Args>(args)...);
 }
 
+/*!
+ * \ingroup DataBoxGroup
+ * \brief Apply the function `F` mutating its `return_tags` and taking as
+ * additional arguments its `argument_tags` and `args`.
+ *
+ * \details
+ * `F` must have `tmpl::list` type aliases `return_tags` and `argument_tags`, as
+ * well as a static `apply` function. The `apply` function must take the types
+ * of the `return_tags` as `gsl::not_null` pointers, followed by the types of
+ * the `argument_tags` as constant references. It can also take the `Args` as
+ * additional arguments.
+ *
+ * \example
+ * \snippet Test_DataBox.cpp mutate_apply_struct_definition_example
+ * This is how to use `mutate_apply` with the above class:
+ * \snippet Test_DataBox.cpp mutate_apply_struct_example_with_tags_lists
+ *
+ * \see box
+ * \tparam F The function to apply
+ * \param box The DataBox out of which to retrieve the Tags to pass to `F`
+ * \param args The arguments to pass to the function that are not in the
+ * DataBox, `box`
+ */
+template <typename F, typename BoxTags, typename... Args>
+inline constexpr auto mutate_apply(const gsl::not_null<DataBox<BoxTags>*> box,
```

Comment:
The other `mutate_apply` can take an object with a non-static apply, and it would be nice if that worked here too.  It looks like that would make the overloads ambiguous, though, so I'm not sure what to do about that.

Reply:
I considered writing some meta programming code for that other overload to check if the type `F` exposes the tags typelists. But I have no immediate use case and so I didn’t bother. Could probably be done at some point though.

Reply:
I think adding that functionality to the current design may not be possible because of the restriction on partial specializations of functions, so I think we need to consider it in this initial design.

Reply:
That functionality would be for the other overload, the one that takes an instance of `F`. This new overload is just an addition.

Reply:
I don't understand.  Taking an instance of `F` and reading tags from `F` seem orthogonal to me.  Why would we want to support two ways to get the tags for a static apply but only one for a call operator (for a non-default-constructed object)?

Reply:
In this PR I'm not changing the current `mutate_apply` function. I only add a more convenient overload for it. I'm not sure what you are asking me to do here.

Reply:
The specific thing I think needs discussion is that it may be impossible to extend this design to match an aspect of the functionality of the original version that is unrelated to the new feature.  Maybe we are OK with that, but this isn't a question we can put off to an issue or a later PR.

So is losing this functionality acceptable?  It is mostly used with lambdas, which are irrelevant here, so maybe it is.  On the other hand, functors with parameters or state are not uncommon.  If losing it is not acceptable, then how do we adjust the design to keep it?

Reply:
This new overload won't take a functor with state, it is designed to take an explicit template parameter with a static function and typelists for arguments and return tags. We have the other (existing) overload for functors with state, and I don't intend to remove it. If we want to make that other overload also read the typelists when they are available, that's a separate issue but could probably be accomplished.

I feel like we are repeating ourselves here, so if this is still not resolved let's talk in a hangout.

Reply:
> If we want to make that other overload also read the typelists when they are available, that's a separate issue but could probably be accomplished.

How?  I don't think that can be accomplished.  This is what I am trying to say.

We can talk in a hangout if you want, but I don't know when I'm going to be available for such a discussion.

Reply:
If you don’t know when you are available to talk, then please make an actionable suggestion so that we can move on

Reply:
Do one of:
* Propose a method by which this feature could be extended to the instance case.  (Does not have to be an implementation, just an explanation of how it is possible.  The implementation can be done later if we decide we want it.)
* Provide an argument for only allowing this feature on the non-instance version.
* Alter the design so that it can be extended to the instance version.

Reply:
@nilsdeppe and myself implemented something, please check the fixup and see if that resolves your issue